src/Orleans/Statistics/CounterStatistic.cs

```diff
@@ -179,11 +185,27 @@ public void IncrementBy(long n)
         {
             if (isOrleansManagedThread)
             {
-                while (perOrleansThreadCounters.Count <= id)
+                ReferenceLong orleansValue = null;
+               
+                try
```

Comment:
I would re-write this with an explicit check for id vs allValuesFromSpecificThread.Count before the access and growing the allValuesFromSpecificThread there instead of exception.
In general, in .NET guidelines, exceptions are for error cases while here it is a legit control flow.
This will also allow to collapse ` if (orleansValue == null)` check in the growing case.